File: binding/include/libqfileiconprovider.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qfileiconprovider.html
pub const qfileiconprovider = struct {

    /// New constructs a new QFileIconProvider object.
    pub fn New() QtC.QFileIconProvider {
        return qtc.QFileIconProvider_new();
    }


    pub fn Icon(self: ?*anyopaque, typeVal: i32) QtC.QIcon {
        return qtc.QFileIconProvider_Icon(@ptrCast(self), @intCast(typeVal));
    }

    /// Allows for overriding the related default method
    pub fn OnIcon(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) QtC.QIcon) void {
        qtc.QFileIconProvider_OnIcon(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseIcon(self: ?*anyopaque, typeVal: i32) QtC.QIcon {
        return qtc.QFileIconProvider_QBaseIcon(@ptrCast(self), @intCast(typeVal));
    }

    pub fn Icon2(self: ?*anyopaque, info: ?*anyopaque) QtC.QIcon {
        return qtc.QFileIconProvider_Icon2(@ptrCast(self), @ptrCast(info));
    }

    /// Allows for overriding the related default method
    pub fn OnIcon2(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) QtC.QIcon) void {
        qtc.QFileIconProvider_OnIcon2(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseIcon2(self: ?*anyopaque, info: ?*anyopaque) QtC.QIcon {
        return qtc.QFileIconProvider_QBaseIcon2(@ptrCast(self), @ptrCast(info));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QFileIconProvider_Delete(@ptrCast(self));
    }
};

File: binding/src/libqfileiconprovider.cpp
#include <QAbstractFileIconProvider>
#include <QFileIconProvider>
#include <QFileInfo>
#include <QIcon>
#include <qfileiconprovider.h>
#include "libqfileiconprovider.h"
#include "libqfileiconprovider.hxx"

QFileIconProvider* QFileIconProvider_new() {
	 return new VirtualQFileIconProvider();
}

QIcon* QFileIconProvider_Icon(const QFileIconProvider* self, int typeVal) {
	return new QIcon(self->icon(static_cast<QAbstractFileIconProvider::IconType>(typeVal)));
}

QIcon* QFileIconProvider_Icon2(const QFileIconProvider* self, const QFileInfo* info) {
	return new QIcon(self->icon(*info));
}

// Base class handler implementation
QIcon* QFileIconProvider_QBaseIcon(const QFileIconProvider* self, int typeVal) {
	auto* vqfileiconprovider = dynamic_cast<const VirtualQFileIconProvider*>(self);
	if (vqfileiconprovider && vqfileiconprovider->isVirtualQFileIconProvider) {
vqfileiconprovider->setQFileIconProvider_Icon_IsBase(true);
	return new QIcon(vqfileiconprovider->icon(static_cast<QAbstractFileIconProvider::IconType>(typeVal)));
}
}

// Auxiliary method to allow providing re-implementation
void QFileIconProvider_OnIcon(const QFileIconProvider* self, intptr_t slot) {
	auto* vqfileiconprovider = dynamic_cast<const VirtualQFileIconProvider*>(self);
	if (vqfileiconprovider && vqfileiconprovider->isVirtualQFileIconProvider) {
vqfileiconprovider->setQFileIconProvider_Icon_Callback(reinterpret_cast<VirtualQFileIconProvider::QFileIconProvider_Icon_Callback>(slot));
}
}

// Base class handler implementation
QIcon* QFileIconProvider_QBaseIcon2(const QFileIconProvider* self, const QFileInfo* info) {
	auto* vqfileiconprovider = dynamic_cast<const VirtualQFileIconProvider*>(self);
	if (vqfileiconprovider && vqfileiconprovider->isVirtualQFileIconProvider) {
vqfileiconprovider->setQFileIconProvider_Icon2_IsBase(true);
	return new QIcon(vqfileiconprovider->icon(*info));
}
}

// Auxiliary method to allow providing re-implementation
void QFileIconProvider_OnIcon2(const QFileIconProvider* self, intptr_t slot) {
	auto* vqfileiconprovider = dynamic_cast<const VirtualQFileIconProvider*>(self);
	if (vqfileiconprovider && vqfileiconprovider->isVirtualQFileIconProvider) {
vqfileiconprovider->setQFileIconProvider_Icon2_Callback(reinterpret_cast<VirtualQFileIconProvider::QFileIconProvider_Icon2_Callback>(slot));
}
}

void QFileIconProvider_Delete(QFileIconProvider* self) {
    delete self;
}

